### PR TITLE
GH-2452 shacl language subtag

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #2452 

Briefly describe the changes proposed in this PR:

Use the built in support for BCP 47 in Java to handle language tags in the SHACL engine for `sh:languageIn`.

---- 
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by hmottestad** [comment]
@rdstn this should fix your issue

**Comment by hmottestad** [comment]
TODO: Retarget against master.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java [hunk 1]
```diff
@@ -33,12 +40,31 @@ boolean checkTuple(Tuple t) {
...
+		// early matching
+		boolean languageMatches = language.map(String::toLowerCase).filter(normalizedLanguageIn::contains).isPresent();
+		if (languageMatches) {
+			return true;
+		}
+
+		// test according to BCP47
+		String langTag = language.get();
+
+		return checkBCP47CompliantMatch(languageRanges, langTag);
```
Comment (jeenbroekstra) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java:56]: I think you can DRY this all this up quite a bit by reusing `Literals.isValidLanguageTag`, or `Literals.normalizeLanguageTag` perhaps.

**Review by jeenbroekstra — CHANGES_REQUESTED** [review]

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java (hunk 1 above)
Comment (hmottestad) [reply core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java:56]: The normalize language tag might help.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java [hunk 2]
```diff
@@ -33,12 +40,31 @@ boolean checkTuple(Tuple t) {
...
 
 	}
 
+	private static boolean checkBCP47CompliantMatch(List<Locale.LanguageRange> languageRanges, String langTag) {
+		List<String> strings = Locale.filterTags(languageRanges, Collections.singletonList(langTag));
```
Comment (hmottestad) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java:61]: This needs to use basic language filtering.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java (hunk 1 above)
Comment (hmottestad) [reply core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/planNodes/LanguageInFilter.java]: I've ended up just using lowercase for the initial fast language matching and using the SPARQL langMatch code from the StrictEvaluationFactory.

File: core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java [hunk 3]
```diff
@@ -1432,6 +1423,20 @@ public Value evaluate(LangMatches node, BindingSet bindings)
...
+		} else if (langTag.length() == langRange.length()) {
+			result = langTag.equalsIgnoreCase(langRange);
+		} else if (langTag.length() > langRange.length()) {
+			// check if the range is a prefix of the tag
+			String prefix = langTag.substring(0, langRange.length());
+			result = prefix.equalsIgnoreCase(langRange) && langTag.charAt(langRange.length()) == '-';
+		}
+		return result;
+	}
+
```
Comment (jeenbroekstra) [inline core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java:1439]: Good find that we are reusing this, but I don't think is the right place for this static method. I suggest we move this to the `Literals` utility class (which has other language-tag related utility methods as well, so it's a natural spot).

**Review by jeenbroekstra — COMMENTED** [review]
Mostly looks good, just one suggestion to move the public static method to a better place.

**Pull request merged**